src/msg.zig

```diff
@@ -0,0 +1,171 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+const std = @import("std");
+
+// pub const MaxStdOutSize = 512; // ensure debug msg are not too long
+
+/// MsgBuffer return messages from a raw text read stream,
+/// according to the following format `<msg_size>:<msg>`.
+/// It handles both:
+/// - combined messages in one read
+/// - single message in several read (multipart)
+/// It is safe (and good practice) to reuse the same MsgBuffer
+/// on several reads of the same stream.
+pub const MsgBuffer = struct {
+    size: usize = 0,
+    buf: []u8,
+    pos: usize = 0,
+
+    pub fn init(alloc: std.mem.Allocator, size: usize) std.mem.Allocator.Error!MsgBuffer {
+        const buf = try alloc.alloc(u8, size);
+        return .{ .buf = buf };
+    }
+
+    pub fn deinit(self: MsgBuffer, alloc: std.mem.Allocator) void {
+        alloc.free(self.buf);
+    }
+
+    fn isFinished(self: *MsgBuffer) bool {
+        return self.pos >= self.size;
+    }
+
+    fn isEmpty(self: MsgBuffer) bool {
+        return self.size == 0 and self.pos == 0;
+    }
+
+    fn reset(self: *MsgBuffer) void {
+        self.size = 0;
+        self.pos = 0;
+    }
+
+    // read input
+    // - `do_func` is a callback to execute on each message of the input
+    // - `data` is a arbitrary payload that will be passed to the callback along with
+    // the message itself
+    pub fn read(
+        self: *MsgBuffer,
+        alloc: std.mem.Allocator,
+        input: []const u8,
+        data: anytype,
+        comptime do_func: fn (data: @TypeOf(data), msg: []const u8) anyerror!void,
+    ) !void {
+        var _input = input; // make input writable
+
+        while (true) {
+            var msg: []const u8 = undefined;
+
+            // msg size
+            var msg_size: usize = undefined;
+            if (self.isEmpty()) {
+                // parse msg size metadata
+                const size_pos = std.mem.indexOfScalar(u8, _input, ':').?;
+                const size_str = _input[0..size_pos];
+                msg_size = try std.fmt.parseInt(u32, size_str, 10);
+                _input = _input[size_pos + 1 ..];
+            } else {
+                msg_size = self.size;
+            }
+
+            // multipart
+            const is_multipart = !self.isEmpty() or _input.len < msg_size;
+            if (is_multipart) {
+
+                // set msg size on empty MsgBuffer
+                if (self.isEmpty()) {
+                    self.size = msg_size;
+                }
+
+                // get the new position of the cursor
+                const new_pos = self.pos + _input.len;
+
+                // check if the current input can fit in MsgBuffer
+                if (new_pos > self.buf.len) {
+                    // max_size is the max between msg size and current new cursor position
+                    const max_size = @max(self.size, new_pos);
```

Comment:
No because `realloc` replace the old size, it's not adding it.
But there is a bug, I sould use `@max(self.buf.len + self.size, new_pos)`
I've changed the name of the variable to be more explanatory (new_size) and updated the comment:

                    // we want to realloc at least:
                    // - a size big enough to fit the entire input (ie. new_pos)
                    // - a size big enough (ie. current size + starting size)
                    // to avoid multiple reallocation